### PR TITLE
Cryopods use signals to handle objectives and other objective improvements

### LINKED ISSUE #5116: Mark as complete doesn't work
I marked AI stay alive until the end as complete and it wasn't completed

### PR DESCRIPTION
**Pull request by chloecat0**
## About The Pull Request
Instead of looping through every objective whenever someone cryos, a signal will be sent to any objectives that have that person as a target. 
Makes find_target actually check is_valid_target so ipcs, etc won't be selected for certain objectives. 
Fixes some issues with cult such as runtimes when making sacrifice image and being impossible to win if no target is set. 
Obsessed objectives won't loop through every job for every player anymore and a new obsession will be found if the previous cryos.
Makes team and crew objectives show up when announcing objectives.
Fixes crew objectives not being added to people at roundstart.
All objectives will check if they're marked as complete. Fixes #5116 

Probably needs to be testmerged.

## Why It's Good For The Game
Less shitcode, better performance, less bugs (hopefully).

## Changelog
:cl:
fix: Fixed crew objectives not being assigned at roundstart
refactor: Cryopods use signals to handle objectives instead of looping through every objective
code: Many minor improvements to objectives
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
Why wasnt this done when DCS was added S M H

File: code/__HELPERS/roundend.dm [hunk 1]
```diff
@@ -217,7 +217,7 @@
 			if(CONFIG_GET(flag/allow_crew_objectives))
 				var/mob/M = C?.mob
 				if(M?.mind?.current && LAZYLEN(M.mind.crew_objectives))
-					for(var/datum/objective/crew/CO in M.mind.crew_objectives)
+					for(var/datum/objective/CO in M.mind.crew_objectives)
```
Comment (Archanial) [inline code/__HELPERS/roundend.dm]: What's the reason of this change? It's supposed to only handle crew objectives, which are different.

File: code/controllers/subsystem/job.dm [hunk 2]
```diff
@@ -26,6 +26,10 @@ SUBSYSTEM_DEF(job)
 		"Chief Medical Officer" = 5,
 		"Head of Security" = 6)
 
+	//Crew Objective stuff
+	var/list/crewobjlist = list()
+	var/list/crewobjjobs = list()
```
Comment (Archanial) [inline code/controllers/subsystem/job.dm]:
```suggestion
	var/list/crew_obj_list = list()
	var/list/crew_obj_jobs = list()
```

File: code/controllers/subsystem/job.dm [hunk 3]
```diff
@@ -37,6 +41,12 @@ SUBSYSTEM_DEF(job)
 
 	spare_id_safe_code = "[rand(0,9)][rand(0,9)][rand(0,9)][rand(0,9)][rand(0,9)]"
 
+	crewobjlist = subtypesof(/datum/objective/crew)
+	for(var/datum/objective/crew/obj as() in crewobjlist) //taken from old Hippie's "job2obj" proc with adjustments.
+		var/list/availableto = splittext(initial(obj.jobs),",")
+		for(var/job in availableto)
+			crewobjjobs["[job]"] += list(obj)
```
Comment (Archanial) [inline code/controllers/subsystem/job.dm]:
```suggestion
			crewobjjobs["[job]"] += obj
```

File: code/datums/brain_damage/creepy_trauma.dm [hunk 4]
```diff
@@ -77,6 +79,23 @@
 	if(hugged == obsession)
 		obsession_hug_count++
 
+/datum/brain_trauma/special/obsessed/proc/on_obsession_cryoed()
+	SIGNAL_HANDLER
+
+	UnregisterSignal(obsession.mind, COMSIG_MIND_CRYOED)
```
Comment (Archanial) [inline code/datums/brain_damage/creepy_trauma.dm:87]: unregister signal on removal of datum

File: code/datums/mind.dm [hunk 5]
```diff
@@ -402,10 +400,14 @@
 				for(var/datum/mind/M in other_owners)
 					output += "<li>Conspirator: [M.name]</li>"
 				output += "</ul>"
+	if(crew_objectives.len)
+		output += "<br><B>Optional Objectives:</B>"
+		for(var/datum/objective/objective in crew_objectives)
```
Comment (Archanial) [inline code/datums/mind.dm]:
```suggestion
		for(var/datum/objective/objective as() in crew_objectives)
```

File: code/datums/mind.dm [hunk 6]
```diff
@@ -590,15 +592,23 @@
...
 	var/obj_count = 1
-	to_chat(current, "<span class='notice'>Your current objectives:</span>")
-	for(var/objective in get_all_objectives())
-		var/datum/objective/O = objective
-		to_chat(current, "<B>Objective #[obj_count]</B>: [O.explanation_text]")
-		obj_count++
+	var/list/all_objectives = get_all_objectives()
+	if(all_objectives.len)
+		to_chat(current, "<span class='notice'>Your current objectives:</span>")
+		for(var/datum/objective/O in all_objectives)
```
Comment (Archanial) [inline code/datums/mind.dm]:
```suggestion
		for(var/datum/objective/O as() in all_objectives)
```

File: code/datums/mind.dm [hunk 7]
```diff
@@ -590,15 +592,23 @@
...
+			to_chat(current, "<B>Objective #[obj_count]</B>: [O.explanation_text]")
+			obj_count++
+	if(crew_objectives.len)
+		to_chat(current, "<span class='notice'>Your optional objectives:</span>")
+		for(var/datum/objective/C in crew_objectives)
```
Comment (Archanial) [inline code/datums/mind.dm]:
```suggestion
		for(var/datum/objective/C as() in crew_objectives)
```

File: code/datums/mind.dm [hunk 8]
```diff
@@ -590,15 +592,23 @@
...
 
 /datum/mind/proc/announce_objectives()
 	var/obj_count = 1
-	to_chat(current, "<span class='notice'>Your current objectives:</span>")
-	for(var/objective in get_all_objectives())
-		var/datum/objective/O = objective
-		to_chat(current, "<B>Objective #[obj_count]</B>: [O.explanation_text]")
-		obj_count++
+	var/list/all_objectives = get_all_objectives()
+	if(all_objectives.len)
```
Comment (Archanial) [inline code/datums/mind.dm]: crew objectives aren't included here?

File: code/__HELPERS/roundend.dm (hunk 1 above)
Comment (chloecat0) [reply code/__HELPERS/roundend.dm]: There's no reason an admin couldn't add a normal objective to someone's crew objective list, this way it'd show up as completed.

File: code/datums/mind.dm (hunk 8 above)
Comment (chloecat0) [reply code/datums/mind.dm]: I decided to keep them separate so they show up as optional and aren't numbered. I'll change the names so they make more sense.

File: code/__HELPERS/roundend.dm (hunk 1 above)
Comment (Archanial) [reply code/__HELPERS/roundend.dm]: If there is no reason, revert this change. Also add as() since you don't have to check types here.

Comment (chloecat0) [reply code/__HELPERS/roundend.dm]: Okay done

File: code/game/gamemodes/objective.dm [hunk 9]
```diff
@@ -276,6 +307,16 @@ GLOBAL_LIST_EMPTY(objectives)
...
 		explanation_text = "Free Objective"
 
+/datum/objective/mutiny/on_target_cryo()
+	set_target(null)
+	team.objectives -= src
+	for(var/datum/mind/M as() in team.members)
+		var/datum/antagonist/rev/R = M.has_antag_datum(/datum/antagonist/rev)
+		R?.objectives -= src
+		to_chat(M.current, "<BR><span class='userdanger'>Your target is no longer within reach. Objective removed!</span>")
+		M.announce_objectives()
```
Comment (KubeRoot) [inline code/game/gamemodes/objective.dm]: If you're using conditional `R?.objectives -= src` here, why not do a proper `if` to also only print things when the antag datum is found?

File: code/game/gamemodes/objective.dm [hunk 10]
```diff
@@ -342,14 +382,13 @@ GLOBAL_LIST_EMPTY(objectives)
...
-	return target
 
 /datum/objective/protect/check_completion()
 	var/obj/item/organ/brain/brain_target
 	if(human_check)
-		brain_target = target.current.getorganslot(ORGAN_SLOT_BRAIN)
+		brain_target = target?.current.getorganslot(ORGAN_SLOT_BRAIN)
 	//Protect will always suceed when someone suicides
-	return !target || considered_alive(target, enforce_human = human_check) || (human_check == TRUE && brain_target)? brain_target.suicided : FALSE
+	return ..() || !target || considered_alive(target, enforce_human = human_check) || (human_check == TRUE && brain_target)? brain_target?.suicided : FALSE
```
Comment (KubeRoot) [inline code/game/gamemodes/objective.dm]:
```suggestion
	return ..() || !target || considered_alive(target, enforce_human = human_check) || (human_check == TRUE && brain_target)? brain_target.suicided : FALSE
```

The null check shouldn't be necessary here, as you check `brain_target` just before in the `&&`.

**Review by KubeRoot — CHANGES_REQUESTED** [review]
Got a couple code nitpicks, looks good in general.

**Review by KubeRoot — APPROVED** [review]

**Pull request merged**